### PR TITLE
Add support ValueObject for Comparison filter

### PR DESCRIPTION
**Pull request by peter-gribanov**
Fix for #133

[ValueObject](https://en.wikipedia.org/wiki/Value_object) must have `__toString()` method for convert it to query parameter.

Example usage:

```php
use MabeEnum\Enum;

class UserStatus extends Enum
{
    const INACTIVE = 'i';
    const ACTIVE   = 'a';
    const DELETED  = 'd';
}
```

In spec 

```php
Spec::eq('status', UserStatus::ACTIVE());
```

**PS:** I not add tests for this solution becos i don't know how to use `phpspec`. Sorry

### CODE REVIEW

**Comment by peter-gribanov** [comment]
This is simple solution for #133

For full support [Doctrine Custom Mapping Types](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/cookbook/custom-mapping-types.html) need use origin type convertor. For this need to know [the type name](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Connection.php#L1432) that can convert this value.

I change code for get type name from class name and try get custom type. 

An alternative solution is to go through the entire [list of registered types](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Types/Type.php#L70) and check each to support this value.

File: src/Filter/Comparison.php [hunk 1]
```diff
@@ -91,7 +91,19 @@ public function getFilter(QueryBuilder $qb, $dqlAlias)
         }
 
         $paramName = $this->getParameterName($qb);
-        $qb->setParameter($paramName, $this->value);
+        $value = $this->value;
+
+        if (is_object($this->value)) { // maybe it's a ValueObject
```
Comment (Nyholm) [inline src/Filter/Comparison.php]: Sorry for my ignorance. Am I reading the docs wrong if I suggest this?
```php
if ($this->value instanceof Type) {
  $value = $this->value->convertToDatabaseValue($value, $qb->getEntityManager()->getConnection()->getDatabasePlatform())
}
```

File: src/Filter/Comparison.php [hunk 2]
```diff
@@ -91,7 +91,19 @@ public function getFilter(QueryBuilder $qb, $dqlAlias)
...
+            // try get type name from class name
+            $class_name_parts = explode('\\', get_class($this->value));
```
Comment (Nyholm) [inline src/Filter/Comparison.php]: We should use camelCase instead of snake_case

File: src/Filter/Comparison.php (hunk 1 above)
Comment (peter-gribanov) [inline src/Filter/Comparison.php]: You are wrong.
Doctrine Mapping Types is not related with Value Object.

To better understand how this works, it's best to look at a specific implementation.

You can see how it works in my library of implemented Interval [Value Object](https://github.com/gpslab/interval).

[This](https://github.com/gpslab/interval/blob/master/src/DateTime/DateTimeInterval.php) is implementation of DateTime Interval Value Object
And [this](https://github.com/gpslab/interval/blob/master/src/Persistence/Doctrine/DBAL/Types/DateTimeIntervalType.php) is a Doctrine Mapping Types for persist DateTime Interval in Doctrine.
If you look, you will see that they are not related.

Example Doctrine [configureation](https://github.com/gpslab/interval#persistence-in-doctrine).

Comment (Nyholm) [inline src/Filter/Comparison.php]: Okey, thanks. I stand corrected.

**Comment by Nyholm** [comment]
Thank you for this PR. Have a look at the style fixes and I'll be happy to merge.

File: src/Filter/Comparison.php (hunk 2 above)
Comment (peter-gribanov) [inline src/Filter/Comparison.php]: fixid

**Comment by peter-gribanov** [comment]
@Nyholm Can, the conversion of the value must also be added to the filter [`In`](https://github.com/Happyr/Doctrine-Specification/blob/master/src/Filter/In.php#L51)?

**Comment by Nyholm** [comment]
Im not sure I understand.

**Comment by peter-gribanov** [comment]
Example usage:

```php
Spec::in('status', [UserStatus::INACTIVE(), UserStatus::DELETED()]);
```

In this case as value indicated list of Value Objects

**Comment by Nyholm** [comment]
Yes, we should consider that use case as well.

**Comment by peter-gribanov** [comment]
I added separate classes for convert value to database value

File: src/DBALTypesResolver.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+     *
+     * @return Type|null
+     */
+    public static function tryGetTypeForValue($value)
+    {
+        if (is_object($value)) { // maybe it's a ValueObject
+            // use class name as type name
+            $classNameParts = explode('\\', get_class($value));
+            $typeName = array_pop($classNameParts);
+
```
Comment (peter-gribanov) [inline src/DBALTypesResolver.php]: Here you can add other rules to find the type

**Comment by peter-gribanov** [comment]
Now, if impossible automatically resolve name of the mapping type by Value Object user can yourself register custom type:

```php
\Doctrine\DBAL\Types\Type::addType('user_status', UserStatusType::class);
\Happyr\DoctrineSpecification\DBALTypesResolver::addType('user_status', UserStatus::class);
```
and then we can get mapping type by Value Object
```php
// return UserStatusType custom type
$type = \Happyr\DoctrineSpecification\DBALTypesResolver::tryGetTypeForValue(UserStatus::ACTIVE());
```

File: src/DBALTypesResolver.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+use Doctrine\DBAL\Types\Type;
+
+abstract class DBALTypesResolver
+{
+    /**
+     * The map of supported doctrine mapping types.
+     *
+     * @var array
+     */
+    private static $_typesMap = array(
```
Comment (Nyholm) [inline src/DBALTypesResolver.php]: No need for underscore here.

File: src/DBALTypesResolver.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    );
```
Comment (Nyholm) [inline src/DBALTypesResolver.php]: Put this on the same row

File: src/Filter/In.php [hunk 6]
```diff
@@ -47,8 +48,17 @@ public function getFilter(QueryBuilder $qb, $dqlAlias)
             $dqlAlias = $this->dqlAlias;
         }
 
+        $value = $this->value;
+        if (is_array($value)) {
```
Comment (Nyholm) [inline src/Filter/In.php]: Isn't this always an array?

File: src/DBALTypesResolver.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+     *
+     * If type is not found return NULL.
+     *
+     * @param mixed $value
+     *
+     * @return Type|null
+     */
+    public static function tryGetTypeForValue($value)
+    {
+        if (is_object($value)) { // maybe it's a ValueObject
```
Comment (Nyholm) [inline src/DBALTypesResolver.php]: If you do if(!is_ob... , it will improve readability.

File: src/DBALTypesResolver.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace Happyr\DoctrineSpecification;
+
+use Doctrine\DBAL\Types\Type;
+
+abstract class DBALTypesResolver
```
Comment (Nyholm) [inline src/DBALTypesResolver.php]: Why is it declared abstract? Do you mean final? Same with the value converter.

File: src/ValueConverter.php [hunk 9]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Happyr\DoctrineSpecification;
+
+use Doctrine\ORM\QueryBuilder;
+
+abstract class ValueConverter
```
Comment (Nyholm) [inline src/ValueConverter.php]: What do you think about putting the logic from the DBALTypesResolver I this class?

File: src/DBALTypesResolver.php (hunk 8 above)
Comment (peter-gribanov) [inline src/DBALTypesResolver.php]: Yes. I mean final. Fix it

File: src/Filter/In.php (hunk 6 above)
Comment (peter-gribanov) [inline src/Filter/In.php]: No. It can be single value. It's works now

File: src/ValueConverter.php (hunk 9 above)
Comment (peter-gribanov) [inline src/ValueConverter.php]: I thought about it. The objectives of these classes are different. I can combine these classes if you think fit

File: src/DBALTypesResolver.php (hunk 4 above)
Comment (peter-gribanov) [inline src/DBALTypesResolver.php]: fixed

File: src/DBALTypesResolver.php (hunk 5 above)
Comment (peter-gribanov) [inline src/DBALTypesResolver.php]: fixed

File: src/DBALTypesResolver.php (hunk 7 above)
Comment (peter-gribanov) [inline src/DBALTypesResolver.php]: fixed

Comment (Nyholm) [inline src/DBALTypesResolver.php]: Great. Thanks!

**Pull request merged**